protocol/l1-upgrades.md

```diff
@@ -0,0 +1,338 @@
+   guards.
+
+More over, it does not present a plan for managing contracts which are not proxied, and not
+Multichain Prepared (MCP-L1) contracts (ie. FaultDisputeGame, AnchorStateRegistry). This
+work will either include, or else depend on, all L1 contracts becoming either:
+
+1. Proxied and MCP ready
+2. Superchain-shared contracts (proxied or not)
+
+# Summary
+
+<!-- Most (if not all) documents should have a summary.
+While the length will likely be proportional to the length of the full document,
+the summary should be as succinct as possible. -->
+
+The proposed solution will perform upgrades by transferring upgrade authorization of all L1
+contracts to the release specific OPCM. Upon completion of the upgrade, this authorization will
+be returned to the Security Council.
+
+Rather than the curent `StorageSetter` based approach which simply deletes the `_initialized` slot,
+allowing for `initialize()` to be called, we take this two-step upgrade further, with a short-lived
+`Populator` contract being temporarily set as the implementation in order to set or modify any
+values as required. This has two significant benefits. First it removes the need to keep a one-time
+`initialize()` function in the run-time logic. Secondly it enables a clean separation of concerns
+between contracts which are newly deployed (requiring all values to be set), and contracts being
+upgraded (requiring only new values to be set).
+
+# Problem Statement + Context
+
+Our upgrades have the following issues:
+
+1. They are divorced from the implementation, meaning the work of writing the upgrade script occurs
+   in `superchain-ops` apart from the work of writing the updated code in the `optimism` monorepo.
+2. The act of writing the upgrade script is onerous, requiring:
+   1. A bespoke script with custom assertions.
+   2. A bespoke Validations.md document
+3. The work required to perform the upgrade is difficult for signers, as they must work through the
+   Validations.md document. This work has taken up to an hour in the past.
+4. The effort described above will scale linearly with the number of OP Chains.
+
+# Proposed Solution
+
+Each new release will have its own OP Contracts Manager which can perform the following distinct
+actions:
+
+1. Deploy new proxies for a new OP Chain (this work is already covered by [link to Blaine's doc]())
+1. Upgrade all existing OP Chains.
+1. Deploy any newly added superchain-shared contracts.
+1. Upgrade existing superchain-shared contracts.
+
+Note that the above functionality also makes for a good set of milestones, each of which can and
+should be implemented incrementally.
+
+## Declarative storage management
+
+Taking inspiration from Chugsplash, we wish to separate the concern of what bytecode to run, and
+what to write to storage. However we also wish to avoid using the current `StorageSetter` pattern,
+which simply provides logic to "write anything anywhere" in storage.
```

Comment:
> What is the specific value prop of the Populator pattern? This will result in 2x contracts being deployed. 

IMO this pattern is similar to the `OptimismPortalMigration` you describe [here](https://github.com/ethereum-optimism/design-docs/issues/154), and I'd have used the `Migration` name had I seen that issue before writing this doc. 

The specific value props are: 
1. Storage layout awareness, meaning we can use vanilla solidity to set storage variables. ie. `foo = _foo` rather than `StorageSetter.setUint(bytes32(0), _foo)`, where `0` is a magic number, or you use assembly to get `foo.slot`. 
2. Ability to implement more complex logic (ie. if-branches, assertions and external calls), examples:
	1. an [if-branch](https://github.com/ethereum-optimism/optimism/blob/maur/repeat-isthmus-contracts/packages/contracts-bedrock/src/L1/ResourceMetering.sol#L168-L172) called within `OptimismPortal2.initialize()`.
	2. [safety](https://github.com/ethereum-optimism/optimism/blob/maur/repeat-isthmus-contracts/packages/contracts-bedrock/src/L1/SystemConfig.sol#L548-L564) [assertions](https://github.com/ethereum-optimism/optimism/blob/maur/repeat-isthmus-contracts/packages/contracts-bedrock/src/L1/SystemConfig.sol#L492-L493) called in `SystemConfig.intialize()`. 
	3. an [external call](https://github.com/ethereum-optimism/optimism/blob/maur/repeat-isthmus-contracts/packages/contracts-bedrock/src/L1/SystemConfig.sol#L483-L486) in `SystemConfig.initialize()`.

> We could just use StorageSetter and call it in a particular way for each contract 

I don't think we can due to my [response here](https://github.com/ethereum-optimism/design-docs/pull/159/files/34679b374bbf155a64182e5527b58f56a6c07faf#r1850696184) and above.

Reply:
Resolving as we're no long proposing the `Populator` and just adding `upgrade()` to the impl.